app/code/Magento/ImportExport/Model/Import/Adapter.php

```diff
@@ -58,9 +54,72 @@ public static function factory($type, $directory, $source, $options = null)
      * @param mixed $options OPTIONAL Adapter constructor options
      *
      * @return AbstractSource
+     * @deprecated
+     * @see Adapter::createSourceByPath()
      */
     public static function findAdapterFor($source, $directory, $options = null)
     {
-        return self::factory(pathinfo($source, PATHINFO_EXTENSION), $directory, $source, $options);
+        return self::createBackwardCompatibleInstance()->createSourceByPath($source, $options);
+    }
+
+
+
+    /**
+     * Finds source for import by file path
+     *
+     * @param $path
+     * @param $options
+     *
+     * @return AbstractSource
+     */
+    public function createSourceByPath($path, $options = [])
+    {
+        try {
+            $options = $this->mapBackwardCompatibleFileParserOption($options);
+            $parser = $this->fileSourceFactory->createFromFilePath($path, $options);
+        } catch (UnsupportedPathException $e) {
+            $this->throwUnsupportedFileException($path);
+        } catch (CorruptedFileException $e) {
+            $this->throwUnsupportedFileException($path);
+        }
+
+        return $parser;
+    }
+
+
+    private function extractFileExtension($path)
+    {
+        $fileName = basename($path);
+
+        if (strpos($path, '.') === false) {
+            return $fileName;
+        }
+
+        $fileExtension = substr($fileName, strrpos($fileName, '.') + 1);
+        return $fileExtension;
+    }
+
+    private function throwUnsupportedFileException($path)
```

Comment:
it's required for all methods and properties, no matter whether it private or public one